console-subscriber/src/lib.rs

```diff
@@ -225,75 +354,164 @@ where
         if self.waker_callsites.contains(event.metadata()) {
             let at = SystemTime::now();
-            let mut visitor = WakerVisitor { id: None, op: None };
+            let mut visitor = WakerVisitor::default();
             event.record(&mut visitor);
-
-            match visitor {
-                WakerVisitor {
-                    id: Some(id),
-                    op: Some(op),
-                } => {
-                    self.send(Event::Waker { id, op, at });
-                }
-                _ => {
-                    tracing::warn!("unknown waker event: {:?}", event);
+            match visitor.result() {
+                Some((id, op)) => self.send(Event::Waker { id, op, at }),
+                None => tracing::warn!("unknown waker event: {:?}", event),
+            }
+        } else if self.resource_op_callsites.contains(event.metadata()) {
+            match ctx.current_span().id() {
+                Some(resource_id) if self.is_id_resource(resource_id, &ctx) => {
+                    let mut resource_op_visitor = ResourceOpVisitor::default();
+                    event.record(&mut resource_op_visitor);
+                    let res = resource_op_visitor.result();
+                    let op_data = match res {
+                        Some(ResourceOpData::Poll { op_name, readiness }) => self
+                            .current_spans
+                            .get()
+                            .and_then(|stack| {
+                                let stack = stack.borrow();
+                                let task_id =
+                                    self.first_entered(&stack, |id| self.is_id_spawned(id, &ctx));
+
+                                let async_op_id =
+                                    self.first_entered(&stack, |id| self.is_id_async_op(id, &ctx));
+
+                                task_id.zip(async_op_id)
+                            })
+                            .map(|(task_id, async_op_id)| {
+                                let op_type = OpType::Poll {
+                                    async_op_id,
+                                    task_id,
+                                    readiness,
+                                };
+
+                                (op_name, op_type)
+                            }),
+                        Some(ResourceOpData::StateUpdate { op_name, attrs }) => {
+                            Some((op_name, OpType::StateUpdate(attrs)))
+                        }
+                        None => None,
+                    };
+
+                    if let Some((op_name, op_type)) = op_data {
+                        let at = SystemTime::now();
+                        self.send(Event::ResourceOp {
+                            metadata,
+                            at,
+                            resource_id: resource_id.clone(),
+                            op_name,
+                            op_type,
+                        });
```

Comment:
nit, take it or leave it: this feels like enough code that i would _probably_ want to stick it in a separate function, if that's practical...